Fixes #548 from now on verification happens always call in other thre…

lukasz-szewc opened the pull request:
Should fix problems with #548 and #601

TimvdLippe commented:
I am going to restart this build a couple of times to test the robustness.

lukasz-szewc commented:
It could take a  while. Hopefully we have loops.

In the meantime try this gist : https://gist.github.com/lukasz-szewc/f1ca9cece0a8997e898e5b6ba1a65f0c

TimvdLippe commented:
Got a failed Travis job: https://travis-ci.org/mockito/mockito/jobs/156890987. The culprit seems to be `org.mockitousage.verification.VerificationAfterDelayTest > shouldVerifyNormallyWithSpecificTimes`

lukasz-szewc commented:
I have not touched `VerificationAfterDelayTest` in this PR...

TimvdLippe commented on src/test/java/org/mockitousage/verification/VerificationWithTimeoutTest.java:
```diff
@@ -161,13 +193,47 @@ public void shouldAllowTimeoutVerificationInOrder() throws Exception {
+            return new Runnable() {
+                @Override
+                public void run() {
+                    countDownLatch.countDown();
```
Shouldn't this statement be moved as last statement? If I read the docs correctly, `countDownLatch.await()` will proceed when all Threads have called `countDownLatch.countDown()`.

lukasz-szewc replied:
Yes and no.  :)

If I move it as a last line in run method then I guarantee that invocation on mock has been performed and in that case any time verification (later in test) is pretty pointless. It would not make a difference if the time slot is 1 millisecond or 1000 milliseconds. 

On the other hand, If I move the countDown before execution on mock (like it is proposed right now), then verification after some time makes sense because we are executing at least two activities in parallel.

bric3 replied:
@lukasz-szewc But the goal of the test is not to test these invocation but to test the timeout verification mode. So when we want to ensure calls have been finished before the verification is virtually OK.

Otherwise that was pretty much the idea I had in mind for timeout tests. And we can have the same kind of tricks for the after timeout.

TimvdLippe commented:
Yes I know, was merely stating that the Travis failure we experienced was a different test. Nonetheless, I think it is best if we aggregrate all test fixes in a PR to study the robustness. Would you be willing to do so?

lukasz-szewc commented:
Yup, another commit with `VerificationAfterDelayTest` changes. Take a look.

TimvdLippe commented:
Great thanks, will be restarting these changes a couple of times too. Really nice to have deterministic results now! :tada:

TimvdLippe commented on src/test/java/org/mockitousage/verification/DelayedExecution.java:
```diff
@@ -0,0 +1,51 @@
+        return new Runnable() {
+            @Override
+            public void run() {
+                countDownLatch.countDown();
```
Adding my previous comment again:

Shouldn't this statement be moved as last statement? If I read the docs correctly, `countDownLatch.await()` will proceed when all Threads have called `countDownLatch.countDown()`.

lukasz-szewc replied:
Yes and no. :)

If I move it as a last line in run method then I guarantee that invocation on mock has been performed and in that case any time verification (later in test) is pretty pointless. It would not make a difference if the time slot is 1 millisecond or 1000 milliseconds.

On the other hand, If I move the countDown before execution on mock (like it is proposed right now), then verification after some time makes sense because we are executing at least two activities in parallel.

bric3 commented:
I'm ok with the change. Even if the fact that async are not completely deterministic, it definitely an improvement, plus the test code is elegant!

bric3 commented on src/test/java/org/mockitousage/verification/DelayedExecution.java:
```diff
@@ -0,0 +1,51 @@
+    private final int delay;
+    private final ArrayList<Character> invocations = new ArrayList<Character>();
+
+    public DelayedExecution(ScheduledExecutorService executor, IMethods mock, int delay) {
```
Could you add the TimeUnit parameter?

lukasz-szewc replied:
I will

bric3 commented:
@lukasz-szewc Also after adding the `TimeUnit` parameter to `DelayedExection` ? Could rebase your commits (`added one missing line` is not really the most expressive message) ;)

ChristianSchwarz commented on src/test/java/org/mockitousage/verification/VerificationWithTimeoutTest.java:
```diff
@@ -137,21 +156,32 @@ public void shouldAllowMixingOnlyWithTimeoutAndFail() throws Exception {
+        delayedExecution.allAsyncCallsStarted();
 
         // then
+        verify(mock, timeout(50)).oneArg('0');
```
@lukasz-szewc 
This line must be removed cause this was not tested in the original test case. For the same reason the call with '2' must be delayed by 20ms.

bric3 replied:
@ChristianSchwarz Good point. I overlooked this one thinking the countdown latch semantic forbid that, but there is two way to handle this 
* it's possible add a new instance of `DelayedExecution` and insert `allAsyncCallsStarted` before the last verify
* tweak the `DelayedExecution` API to be more fine grained

About the missing verify I noted it as well in the original test as well, but it may be de to a human error so I'm ok with the addition.

lukasz-szewc replied:
I will keep original intent of that test. However this particular test is at least weird and I have no idea what this test is actually testing

ChristianSchwarz commented:
I like the idea of starting the threads and await there readiness, this might solve the timing issues. An other option is to write an simplified ScheduledExecutorService with the same behavior as the DelayedExecution of this PR. This way the test can remain untouched and we have still the possibiliy run tasks with individual delay as required by one test.

lukasz-szewc commented:
1. added `TimeUnit` to constructor
2. reworked one of test in `VerificationWithTimeoutTest' mentioned above
3. rebased and pushed with force

take a look.

bric3 commented on src/test/java/org/mockitousage/verification/DelayedExecution.java:
```diff
@@ -0,0 +1,54 @@
+
+    private void sleep() {
+        try {
+            timeUnit.sleep(delay);
```
@lukasz-szewc Actually the idea of passing the TimeUnit is to do a `TimeUnit.toMillis(delay)` here ;)

ChristianSchwarz replied:
@bric3 @lukasz-szewc 
better use `timeUnit.sleep(delay)` here ;-)

bric3 replied:
That works too ;)

bric3 replied:
I think I bugged there it's already `timeUnit.sleep(delay);`

bric3 commented on src/test/java/org/mockitousage/verification/VerificationWithTimeoutTest.java:
```diff
@@ -137,11 +157,14 @@ public void shouldAllowMixingOnlyWithTimeoutAndFail() throws Exception {
-        callAsyncWithDelay(mock, '1', 0, MILLISECONDS);
-        callAsyncWithDelay(mock, '2', 20, MILLISECONDS);
+        delayedExecution.allAsyncCallsStarted();
+        twentyMillisDelayedExecution.allAsyncCallsStarted();
```
@lukasz-szewc I'd rather propose

```java
delayedExecution.allAsyncCallsStarted();

verify(mock, timeout(50)).oneArg('1');
twentyMillisDelayedExecution.allAsyncCallsStarted();
verify(mock, timeout(50)).oneArg('2');
```
